contracts/hydro/src/contract.rs

```diff
@@ -676,135 +675,210 @@ fn vote(
+            PROPS_BY_SCORE.remove(
                 deps.storage,
                 (
                     (round_id, proposal.tranche_id),
                     proposal.power.into(),
                     vote.prop_id,
                 ),
-                &vote.prop_id,
+            );
+
+            remove_many_validator_shares_from_proposal(
+                deps.storage,
+                round_id,
+                vote.prop_id,
+                vote.time_weighted_shares
+                    .iter()
+                    .map(|(k, v)| (k.clone(), *v))
+                    .collect::<Vec<_>>(),
             )?;
-        }
 
-        // Delete vote
-        VOTE_MAP.remove(deps.storage, (round_id, tranche_id, info.sender.clone()));
+            // save the new power into the proposal
+            let total_power = get_total_power_for_proposal(deps.as_ref().storage, vote.prop_id)?;
+            proposal.power = total_power.to_uint_ceil(); // TODO: decide whether we need to round or represent as decimals
 
-        response = response.add_attribute("old_proposal_id", vote.prop_id.to_string());
+            // Save the proposal
+            PROPOSAL_MAP.save(
+                deps.storage,
+                (round_id, tranche_id, vote.prop_id),
+                &proposal,
+            )?;
+
+            // Add proposal's new power in PROPS_BY_SCORE
+            if proposal.power > Uint128::zero() {
+                PROPS_BY_SCORE.save(
+                    deps.storage,
+                    (
+                        (round_id, proposal.tranche_id),
+                        proposal.power.into(),
+                        vote.prop_id,
+                    ),
+                    &vote.prop_id,
+                )?;
+            }
+
+            // Delete vote
+            VOTE_MAP.remove(
+                deps.storage,
+                ((round_id, tranche_id), info.sender.clone(), lock_id),
+            );
+
+            response = response.add_attribute("old_proposal_id", vote.prop_id.to_string());
+        }
     }
 
     let lock_epoch_length = CONSTANTS.load(deps.storage)?.lock_epoch_length;
+    let mut voted_proposals = vec![];
 
-    // Get sender's locked shares for each validator
-    let mut time_weighted_shares_map: HashMap<String, Decimal> = HashMap::new();
-    let locks =
-        LOCKS_MAP
-            .prefix(info.sender.clone())
-            .range(deps.storage, None, None, Order::Ascending);
+    for proposal_to_lockups in proposals_votes {
+        let proposal_id = proposal_to_lockups.proposal_id;
 
-    for lock in locks {
-        let (_, lock_entry) = lock?;
+        // TODO: optimize so that proposal related stores are updated only once
+        // TODO: time_weighted_shares on Vote will have only one item
+        for lock_id in proposal_to_lockups.lock_ids {
+            // Get sender's locked shares for each validator
+            let mut time_weighted_shares_map: HashMap<String, Decimal> = HashMap::new();
```

Comment:
this map seems unnecessary now, because each lock_id will have only shares from one validator in it

Reply:
Done in https://github.com/informalsystems/hydro/pull/157/commits/21e74b10f1675c18a3fa0df0beb1d1c15f0458b7